plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -417,11 +417,92 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
          packed_transaction_ptr  trx;
          next_func_t             next;
       };
+   
       // The queue storing read-only transactions to be executed by read-only threads
-      struct ro_trx_queue_t {
-         std::mutex              mtx;
-         std::deque<ro_trx_t>    queue;
+      class ro_trx_queue_t {
+      public:
+         void push_back(ro_trx_t&& trx) {
+            std::unique_lock<std::mutex> g( mtx );
+            queue.push_back(std::move(trx));
+            notify_waiting(g, false);
+         }
+         
+         void push_front(ro_trx_t&& trx) {
+            std::unique_lock<std::mutex> g( mtx );
+            queue.push_front(std::move(trx));
+            notify_waiting(g, false);
+         }
+
+         bool empty() const {
+            std::lock_guard<std::mutex> g( mtx );
+            return queue.empty();
+         }
+
+         // may wait if the queue is empty, and not all other threads are already waiting.
+         // returns true if a transaction was dequeued and should be executed, or false 
+         // if conditions are met to stop processing transactions. 
+         bool pop_front(ro_trx_t &trx) {
+            std::unique_lock<std::mutex> g( mtx );
+            
+            if (should_exit()) {
+               notify_waiting(g, true);
+               return false;
+            }
+
+            if (queue.empty()) {
+               if (num_waiting + 1 == max_waiting) {
```

Comment:
I think this needs to be refactored so you wait on a predicate on the conditions of the condition variable. Seems to me that condition is non-empty queue, received block, or timer fired. All threads can wait on that condition and be notified to check on a push read only trx (notify one) or timer (notify all).

Reply:
In the following diagram, it shows that when we run out of trx to process in every thread, we switch back to the read window. That is what we did before this PR as well. 
![image](https://user-images.githubusercontent.com/1496907/226724577-92ac16d0-f984-4c00-9981-891bebbaa591.png)